python/hail/history.py

```diff
@@ -0,0 +1,170 @@
+
+        (obj._history
+         .add_method(f.__name__, postnl_args, kwargs_not_default)
+         .write(output_path))
+
+        return result
+    return decorator(_write)
+
+
+def format_args(arg, stmts):
+    if isinstance(arg, list):
+        return [format_args(a, stmts) for a in arg]
+    elif isinstance(arg, tuple):
+        return tuple([format_args(a, stmts) for a in arg])
+    elif isinstance(arg, dict):
+        return {format_args(k, stmts): format_args(v, stmts) for k, v in arg.iteritems()}
+    else:
+        if isinstance(arg, HistoryMixin):
+            h = arg._get_history()
+            stmts += remove_dup_stmts(stmts, h.stmts)
+            return h
+        else:
+            return arg
+
+
+def remove_dup_stmts(stmts1, stmts2):
+    return [s2 for s1, s2 in itertools.izip_longest(stmts1, stmts2) if s2 and s1 != s2]
+
+
+class History(object):
+    def __init__(self, expr="", stmts=[]):
+        self.expr = expr
+        self.stmts = stmts
+
+    def set_varid(self, id):
+        return History(id, self.stmts + ['{} = ({})'.format(id, self.expr)])
+
+    def add_method(self, f_name, args, kwargs, index=None, key_name=None):
+        stmts = list(self.stmts) # make a copy
+        f_args = [repr(format_args(a, stmts)) for a in args] + ["{}={}".format(k, repr(format_args(v, stmts))) for k, v in kwargs.iteritems()]
+        expr = self.expr + ("\n    ." + f_name + "(" + ", ".join(f_args) + ")")
+        if index:
+            expr += "[{}]".format(index)
+        elif key_name:
+            expr += "[{}]".format(repr(key_name))
+        return History(expr, stmts)
+
+    @staticmethod
+    def from_classmethod(cls_name, f_name, args, kwargs):
+        stmts = []
+        f_args = [repr(format_args(a, stmts)) for a in args] + ["{}={}".format(k, repr(format_args(v, stmts))) for k, v in kwargs.iteritems()]
+        expr = cls_name + "." + f_name + "(" + ", ".join(f_args) + ")"
+        return History(expr, stmts)
+
+    @staticmethod
+    def from_init(cls_name, args, kwargs):
+        stmts = []
+        f_args = [repr(format_args(a, stmts)) for a in args] + ["{}={}".format(k, repr(format_args(v, stmts))) for k, v in kwargs.iteritems()]
+        expr = cls_name + "(" + ", ".join(f_args) + ")"
+        return History(expr, stmts)
+
+    def write(self, file):
+        with hadoop_write(file) as f:
+            f.write(str(self) + "\n")
+
+    def __repr__(self):
+        return self.expr
+
+    def __str__(self):
+        now = datetime.datetime.now()
+        history = "# {}\n# version: {}\n\n".format(now.isoformat(), Env._hc.version)
+        for stmt in self.stmts:
+            history += (stmt + "\n\n")
+        history += ("(" + self.expr + ")")
+        try:
+            import autopep8
+            return autopep8.fix_code(history)
+        except:
+            return history
+
+
+class HistoryMixin(object):
+    def __init__(self):
+        self._history = None
+
+    def _set_history(self, history):
```

Comment:
do these set/get methods add anything? They're "private" and seem to just wrap

```python
obj._history = history
```
and
```python
obj._history
```